### ISSUE TITLE
Cannot run ngx.run_worker_thread

### ISSUE THREAD

**Issue by bluetoxin** [Kong/kong#11483]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

3.2.2-3.4.0

### Current Behavior

Starting from OpenResty version 1.21.4.x, the developers have introduced a new API called [ngx.run_worker_thread](https://github.com/openresty/lua-nginx-module#ngxrun_worker_thread). This API enables the execution of Lua functions within separate worker threads. The latest Kong versions (greater than 3.0.0) incorporate OpenResty 1.21.4.1, which includes this feature. However, I'm encountering issues when attempting to utilize this function.
```
 attempt to call field 'run_worker_thread' (a nil value)
```
Furthermore, I'm facing difficulties even with adding the "thread_pool" directive in the Nginx Lua template. The problem is illustrated by the following error message:
```
/usr/local/share/lua/5.1/kong/cmd/start.lua:62: nginx configuration is invalid (exit code 1):
nginx: [emerg] unknown directive "thread_pool" in /usr/local/kong/nginx.conf:23
nginx: configuration file /usr/local/kong/nginx.conf test failed

stack traceback:
	[C]: in function 'assert'
	/usr/local/share/lua/5.1/kong/cmd/start.lua:62: in function 'execute'
	/usr/local/share/lua/5.1/kong/cmd/restart.lua:34: in function 'cmd_exec'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:97: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:97>
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/kong/cmd/init.lua:97: in function </usr/local/share/lua/5.1/kong/cmd/init.lua:54>
	/usr/local/bin/kong:9: in function 'file_gen'
	init_worker_by_lua:47: in function <init_worker_by_lua:45>
	[C]: in function 'xpcall'
	init_worker_by_lua:54: in function <init_worker_by_lua:52>
```
In summary, despite the inclusion of the new [ngx.run_worker_thread](https://github.com/openresty/lua-nginx-module#ngxrun_worker_thread) API in OpenResty 1.21.4.x and its integration into Kong versions beyond 3.0.0, I'm encountering difficulties both in invoking the function and in adding the "thread_pool" directive to the Nginx Lua template. The provided error message further details the issue I'm facing.

**Comment by StarlightIbuki** [comment]
Hi. Thanks for reporting this. Could you run `resty -v` and share the result?

**Comment by bluetoxin** [comment]
```
root@8bfa91ab4e6f:/# resty -v 
resty 0.28
nginx version: openresty/1.21.4.1
built with OpenSSL 3.1.1 30 May 2023
TLS SNI support enabled
configure arguments: --prefix=/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/bazel-out/k8-opt/bin/external/openresty/openresty.build_tmpdir/openresty/nginx --with-cc-opt='-O2 -I/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/bazel-out/k8-opt/bin/external/openresty/openresty.ext_build_deps/pcre/include -I/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/bazel-out/k8-opt/bin/external/openresty/openresty.ext_build_deps/openssl/include -I/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/bazel-out/k8-opt/bin/external/openresty/openresty.ext_build_deps/luajit/include -DNGX_WASM_HOST_PROPERTY_NAMESPACE=kong -I/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/external/wasmtime-x86_64/include' --add-module=../echo-nginx-module-0.62 --add-module=../ngx_lua-0.10.21 --add-module=../headers-more-nginx-module-0.33 --add-module=../ngx_stream_lua-0.0.11 --with-ld-opt='-Wl,-rpath,/usr/local/openresty/luajit/lib -L/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/bazel-out/k8-opt/bin/external/openresty/openresty.ext_build_deps/pcre/lib -L/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/bazel-out/k8-opt/bin/external/openresty/openresty.ext_build_deps/openssl/lib -L/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/bazel-out/k8-opt/bin/external/openresty/openresty.ext_build_deps/luajit/lib -Wl,-rpath,/usr/local/kong/lib -Wl,--allow-multiple-definition -L/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/external/wasmtime-x86_64/lib' --with-pcre-jit --with-http_ssl_module --with-http_sub_module --with-http_realip_module --with-http_stub_status_module --with-http_v2_module --with-stream_realip_module --with-stream_ssl_preread_module --add-module=/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/external/lua-kong-nginx-module --add-module=/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/external/lua-kong-nginx-module/stream --add-module=/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/external/lua-resty-lmdb --add-module=/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/external/lua-resty-events --with-compat --add-dynamic-module=/home/runner/.cache/bazel/_bazel_runner/5f5cb99aeb7e2b4db19484ea5022e949/execroot/kong/external/ngx_wasm_module --with-stream --with-stream_ssl_module
```

**Comment by bluetoxin** [comment]
@StarlightIbuki Hey, I wanted to let you know that I've gone ahead and set up a repository with a Minimum Reproducible Example of the bug we were discussing. It'd be great if you could take a look at it. 
https://github.com/bluetoxin/mre-kong-11483

**Comment by bluetoxin** [comment]
The primary benefit that drew me in is the capability to write sophisticated blocking C API, all while avoiding the hindrance of blocking the main thread during execution.

Our team is currently exploring the prospect of refactoring our extensive Lua codebase into more widely adopted programming languages like Go, Rust, or C++. The ultimate goal is to seamlessly transition to Proxy-Wasm, but complete rewrite is not feasible in one fell swoop. Consequently, we've opted to begin by rewriting a portion of it and then interact with it through ffi until most of our codebase has been successfully migrated. However, our plans have hit a roadblock, as compiled shared libraries presently obstruct the execution flow.

Also, from the OpenResty documentation:
```
CPU bound task, e.g. do md5 calculation
File I/O task
Call os.execute() or blocking C API via ffi
Call external Lua library not based on cosocket or nginx
```

**Comment by StarlightIbuki** [comment]
We had a discussion and decided not to introduce the feature for now.
The `worker_thread` has a lot of overhead which cancels out the benefit. It creates a Lua VM and copies the arguments and the return values back and forth.
If you want to use the feature anyway, you may try to build from source with Bazel. Seemingly another possible solution is to handle parallelism from the other side of FFI (if we do not need the Lua code also to be parallized).

**Comment by StarlightIbuki** [comment]
It seems Github does not want me to remove the label. I'll leave the "pending feedback" label and you can ping me to reopen the ticket if it's closed.

**Comment by bluetoxin** [comment]
Ok, thank you!
